### PR TITLE
[GLUTEN-7475][VL] Add a config to control whether add trim node when CAST from varchar

### PR DESCRIPTION
**Pull request by Henry2SS**
## What changes were proposed in this pull request?

Add a config to control whether a trim node is added in CAST. Because we know that removing the trim node can improve the performance by 1.x times. But we are not sure now whether the results will be consistent with spark after removing it.

(Fixes: \#7475)

## How was this patch tested?

Integration tests passed locally. The performance is 1.x times faster.

### CODE REVIEW

**Comment by PHILO-HE** [comment]
@Henry2SS, thanks for your pr! This piece of scala code is for making Gluten/Velox consistent with Spark. Seems the current Velox code doesn't have the consistent white spaces definition? Could you check further?

**Comment by Henry2SS** [comment]
> @Henry2SS, thanks for your pr! This piece of scala code is for making Gluten/Velox consistent with Spark. Seems the current Velox code doesn't have the consistent white spaces definition? Could you check further?

There is a method called `isUnicodeWhiteSpace` in `velox/functions/lib/string/StringImpl.h`. 

And I noticed that, from https://github.com/facebookincubator/velox/pull/7377 this PR, the author proposed different corner cases for Presto and Spark on CAST, including removeWhiteSpaces.

Unfortunately, I don't have much knowledge about this, please confirm.

**Comment by Henry2SS** [comment]
> @Henry2SS, thanks for your pr! This piece of scala code is for making Gluten/Velox consistent with Spark. Seems the current Velox code doesn't have the consistent white spaces definition? Could you check further?

And do we have some unit-tests to test CAST from VARCHAR in CI?
Is it possible to confirm it by all the tests in CI?

**Comment by xumingming** [comment]
> The performance is 1.x times faster.

@Henry2SS Do you mind show the concrete performance improvement?

**Comment by Henry2SS** [comment]
> > The performance is 1.x times faster.
> 
> @Henry2SS Do you mind share the concrete performance improvement?

My pleasure.

There is a benchmark testing CAST VARCHAR to BIGINT. The sql is like `INSERT INTO tmp_table as select CAST(string_col as BIGINT) from table`. And the table has 2,879,966,589 rows.

We used gperftool to find that there is a function that takes up the most of the execution time in benchmark.

Before removing trim node, the total time of `ProjectExecTransformer` is 17.3 min ,dividing into 515 tasks.
After removing trim node, the total time of `ProjectExecTransformer` is 7.5 min ,dividing into 515 tasks.

So it is more than 1.x times faster.

File: backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxSparkPlanExecApi.scala
Comment (jackylee-ch) [inline backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxSparkPlanExecApi.scala:721]: Have you checked if they have the same sementic between vanilla Spark and velox?

Comment (Yohahaha) [reply backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxSparkPlanExecApi.scala:721]:
```
val expect = Seq(Row(123))
checkAnswer(sql("SELECT cast('\\t\\n\\u001F 123\\u000B' as int);"), expect)
```

this UT works.

**Comment by Yohahaha** [comment]
We can merge this PR when GHA passed and no blocking comments.

File: backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxSparkPlanExecApi.scala
Comment (jackylee-ch) [reply backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxSparkPlanExecApi.scala:721]: Maybe add a config, default false, for it? A little worried about the result diff if we haven't tested the total case between vanilla spark and velox.

Comment (Yohahaha) [reply backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxSparkPlanExecApi.scala:721]: > Maybe add a config, default false, for it? A little worried about the result diff if we haven't tested the total case between vanilla spark and velox.

@Henry2SS would you?

Comment (Henry2SS) [reply backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxSparkPlanExecApi.scala:721]: > > Maybe add a config, default false, for it? A little worried about the result diff if we haven't tested the total case between vanilla spark and velox.
> 
> @Henry2SS would you?

I'll do it later.

Comment (jackylee-ch) [reply backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxSparkPlanExecApi.scala:721]: After looking at the velox code, the conversion from string to int/long is basically similar between [vanilla Spark](https://github.com/apache/spark/blob/master/common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java#L200) and [velox](https://github.com/facebookincubator/velox/blob/main/velox/functions/lib/string/StringImpl.h#L354). But for the conversion from string to float/double, [vanilla Spark](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L390) uses `String.trim` to remove characters smaller than spaces, which is not same as velox, so we still need this extra trim.
A new config, which is default false, is good to me. Thanks @Henry2SS for working on it.

Comment (PHILO-HE) [reply backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxSparkPlanExecApi.scala:721]: @Henry2SS, Spark UT doesn't cover many corner cases used in production workload. I'm afraid users would complain result mismatching issues due to this removement. Agree with @jackylee-ch, it's good to use a config to give the control to users. And maybe at some future time point, we can remove it.

Comment (Henry2SS) [reply backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxSparkPlanExecApi.scala]: > @Henry2SS, Spark UT doesn't cover many corner cases used in production workload. I'm afraid users would complain result mismatching issues due to this removement. Agree with @jackylee-ch, it's good to use a config to give the control to users. And maybe at some future time point, we can remove it.

Done.
And thanks for your advice. @PHILO-HE @jackylee-ch @Yohahaha

File: backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxSparkPlanExecApi.scala
Comment (jackylee-ch) [inline backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxSparkPlanExecApi.scala:744]:
```suggestion
          case StringType if GlutenConfig.getConf.castFromVarcharAddTrimNode =>
              val trimNode = StringTrim(
                c.child,
                Some(
                  Literal(trimWhitespaceStr +
                    trimSpaceSepStr + trimLineSepStr + trimParaSepStr)))
              c.withNewChildren(Seq(trimNode)).asInstanceOf[Cast]
```

File: backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxSparkPlanExecApi.scala
Comment (jackylee-ch) [inline backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxSparkPlanExecApi.scala:732]:
```suggestion
          case StringType if GlutenConfig.getConf.castFromVarcharAddTrimNode =>
              val trimNode = StringTrim(c.child, Some(Literal(trimSpaceStr)))
              c.withNewChildren(Seq(trimNode)).asInstanceOf[Cast]
```

File: backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxSparkPlanExecApi.scala
Comment (Henry2SS) [reply backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxSparkPlanExecApi.scala]: done

File: backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxSparkPlanExecApi.scala
Comment (Henry2SS) [reply backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxSparkPlanExecApi.scala]: done

File: shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala
Comment (Yohahaha) [inline shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala:2137]: I think we can set default value to false, because all GHA has passed and performance boost can not be ignored. If incorrect result issue occurs, we can fix it or make this config more fine-grained. 

Just like scan parquet complex types, we remove force fallback config after all UT passed.

@PHILO-HE @jackylee-ch

Comment (jackylee-ch) [reply shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala:2137]: It's ok to me. And if it's set default to false, we need add this config to `Configuration.md` to make sure users know about it.

Comment (Henry2SS) [reply shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala:2137]: > need add this config to

I'll do this.

Comment (Henry2SS) [reply shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala:2137]: > It's ok to me. And if it's set default to false, we need add this config to `Configuration.md` to make sure users know about it.

I'll do this.

**Review by Yohahaha — APPROVED** [review]
👍

**Review by jackylee-ch — APPROVED** [review]

**Review by PHILO-HE — APPROVED** [review]
Looks good. CI once passed with trim node removed, will merge the pr soon.

**Comment by PHILO-HE** [comment]
@Henry2SS, one code style violation.

```
11:10:48  error file=/home/jenkins/agent/workspace/gluten/gluten-ci/ut-stage-5/shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala message=File line length exceeds 100 characters line=2134
```

**Comment by Yohahaha** [comment]
```
12:35:41  [ERROR] Failed to execute goal com.diffplug.spotless:spotless-maven-plugin:2.27.2:check (spotless-check) on project spark-sql-columnar-shims-common: The following files had format violations:
12:35:41  [ERROR]     src/main/scala/org/apache/gluten/GlutenConfig.scala
12:35:41  [ERROR]         @@ -311,7 +311,7 @@
12:35:41  [ERROR]          
12:35:41  [ERROR]          ··def·veloxBloomFilterMaxNumBits:·Long·=·conf.getConf(COLUMNAR_VELOX_BLOOM_FILTER_MAX_NUM_BITS)
12:35:41  [ERROR]          
12:35:41  [ERROR]         -··def·castFromVarcharAddTrimNode·:·Boolean·=·conf.getConf(CAST_FROM_VARCHAR_ADD_TRIM_NODE)
12:35:41  [ERROR]         +··def·castFromVarcharAddTrimNode:·Boolean·=·conf.getConf(CAST_FROM_VARCHAR_ADD_TRIM_NODE)
12:35:41  [ERROR]          
12:35:41  [ERROR]          ··case·class·ResizeRange(min:·Int,·max:·Int)·{
12:35:41  [ERROR]          ····assert(max·>=·min)
12:35:41  [ERROR]         @@ -2131,9 +2131,10 @@
12:35:41  [ERROR]          ··val·CAST_FROM_VARCHAR_ADD_TRIM_NODE·=
12:35:41  [ERROR]          ····buildConf("spark.gluten.velox.castFromVarcharAddTrimNode")
12:35:41  [ERROR]          ······.internal()
12:35:41  [ERROR]         -······.doc("If·true,·will·add·a·trim·node·"·+
12:35:41  [ERROR]         -········"which·has·the·same·sementic·as·vanilla·Spark·to·CAST-from-varchar."·+
12:35:41  [ERROR]         -········"Otherwise,·do·nothing.")
12:35:41  [ERROR]         +······.doc(
12:35:41  [ERROR]         +········"If·true,·will·add·a·trim·node·"·+
12:35:41  [ERROR]         +··········"which·has·the·same·sementic·as·vanilla·Spark·to·CAST-from-varchar."·+
12:35:41  [ERROR]         +··········"Otherwise,·do·nothing.")
12:35:41  [ERROR]          ······.booleanConf
12:35:41  [ERROR]          ······.createWithDefault(false)
12:35:41  [ERROR]          }
12:35:41  [ERROR] Run 'mvn spotless:apply' to fix these violations.
```

CH CI failed due to style check, try use `mvn spotless:apply -P backends-clickhouse -DskipTests` to fix.

**Comment by Henry2SS** [comment]
> ```
> 12:35:41  [ERROR] Run 'mvn spotless:apply' to fix these violations.
> ```
> 
> CH CI failed due to style check, try use `mvn spotless:apply -P backends-clickhouse -DskipTests` to fix.

done

**Pull request merged**
